wiki.moinmoin

```diff
@@ -380,15 +379,54 @@ RULE:     - the owning team will provide timely, high quality updates for the
 RULE:       security team to sponsor to fix issues in the affected vendored code
 RULE:     - if subsequent uploads add new vendored components or dependencies
 RULE:       these have to be reviewed and agreed by the security team.
+RULE:     - Such updates in the project might be trivial, but imply that a
+RULE:       dependency for e.g. a CVE fix will be moved to a new major version.
+RULE:       Being vendored that does gladly at least not imply incompatibility
+RULE:       issues with other packages or the SRU policy. But it might happen
+RULE:       that this triggers either:
+RULE:       a) The need to adapt the current version of the main package and/or
+RULE:          other vendored dependencies to work with the new dependency
+RULE:       b) The need to backport the fix in the dependency as the main
+RULE:          package will functionally only work well with the older version
+RULE:       c) The need to backport the fix in the dependency, as it would imply
+RULE:          requiring a newer toolchain to be buildable that isn't available
+RULE:          in the target release.
+RULE: - The rust ecosystem currently isn't yet considered stable enough for
+RULE:   classic lib dependencies and transitions in main; therefore the
+RULE:   expectation for those packages is to vendor (and own/test) all
+RULE:   dependencies (except those provided by the rust runtime itself).
+RULE:   This implies that all the rules for vendored builds always
+RULE:   apply to them. In addition
+RULE:   - Rust builds using librust-*-dev packges will populate the attribute
+RULE:     `X-Cargo-Built-Using`, right now this is expected not to be present.
```

Comment:
Please require the toolchain to be fixed to use the Built-Using field in Ubuntu, instead of codifying this other field.

Reply:
The idea was to follow Debian which had a debate in Debian to separate classic `Built-Using` from `X-Cargo-Built-Using`/`XB-X-Cargo-Built-Using`, but if that is the preference that is ok.

I rechecked the current state and it seems we are even ok as-is right now.
It seems that nowadays we always get both values.
Currently dh-cargo generates the same content into both variables:
https://sources.debian.org/src/dh-cargo/28/dh-cargo-built-using/?hl=165#L165
And packages usually reference both in d/control:
https://sources.debian.org/src/rust-bindgen/0.59.1-2/debian/control/?hl=243#L242

That means that we should be able to just change the wording to the more common `Cargo-Built-Using` which also means that tools relying on it like the check for component mismatches would already work.

On the first few rust based MIRs we have to manually check that the assumption that all is properly reported via `Built-Using` is really true , I'll add a statement for that in the MIR Team check section. If we find a technical reason some are missed => bug, if we find common patterns that make them be missed => new rules.

Done: 
- I updated the wording to use Cargo-Built-Using
- added a statement about the preliminary status of the rust rules
- I added a statement to double-check the correctness of rust Cargo-Built-Using

Reply:
Actually, the `Built-Using` field only contains the rustc version, while `XB-Cargo-Built-Using` contains all the crates used. There's a trailing `_x` in the code that's fairly easy to miss, I only went back because I looked at the ripgrep binary:

```
zstd -d < control.tar.zst
./0000755000000000000000000000000014113505654007715 5ustar  rootroot./control0000644000000000000000000000511514113505654011322 0ustar  rootrootPackage: ripgrep
Source: rust-ripgrep
Version: 13.0.0-2
Architecture: amd64
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Debian Rust Maintainers <pkg-rust-maintainers@alioth-lists.debian.net>
Installed-Size: 4155
Depends: libc6 (>= 2.34), libgcc-s1 (>= 4.2)
Built-Using: rustc (= 1.53.0+dfsg1+llvm-4ubuntu1)
Section: utils
Priority: optional
Multi-Arch: allowed
Homepage: https://github.com/BurntSushi/ripgrep
Description: Recursively searches directories for a regex pattern
 ripgrep is a line-oriented search tool that recursively searches your current
 directory for a regex pattern while respecting your gitignore rules and
 automatically skip hidden files/directories (smart filtering) and binary files.
 ripgrep is similar to other popular search tools like The Silver Searcher, ack
 and grep.
 .
 ripgrep is generally faster than both The Silver Searcher and GNU grep. It
 defaults to recursive directory search and won't search files ignored by your
 `.gitignore` files. Use ripgrep if you like speed, filtering by default, fewer
 bugs, and Unicode support.
 .
 On the other hand, if you like multiline search, then ripgrep may not quite
 meet your needs (yet), and it will never support fancy regex features such as
 backreferences or lookaround
X-Cargo-Built-Using: rust-aho-corasick (= 0.7.10-1), rust-atty (= 0.2.14-2), rust-base64 (= 0.13.0-1), rust-bitflags (= 1.2.1-1), rust-bstr (= 0.2.12-1), rust-bytecount (= 0.6.0-1), rust-byteorder (= 1.4.3-2), rust-cfg-if-0.1 (= 0.1.10-2), rust-cfg-if (= 1.0.0-1), rust-clap (= 2.33.3-1), rust-crossbeam-utils (= 0.8.5-1), rust-encoding-rs (= 0.8.22-1), rust-encoding-rs-io (= 0.1.6-2build1), rust-fnv (= 1.0.6-1), rust-globset (= 0.4.8-2), rust-grep-cli (= 0.1.6-2), rust-grep (= 0.2.8-2), rust-grep-matcher (= 0.1.5-2), rust-grep-pcre2 (= 0.1.5-1), rust-grep-printer (= 0.1.6-1), rust-grep-regex (= 0.1.9-2), rust-grep-searcher (= 0.1.8-2), rust-ignore (= 0.4.18-2), rust-itoa (= 0.4.3-1), rust-lazy-static (= 1.4.0-1), rust-libc (= 0.2.103-1), rust-log (= 0.4.11-2), rust-memchr (= 2.3.3-1), rust-memmap (= 0.7.0-1), rust-num-cpus (= 1.13.0-1), rust-once-cell (= 1.5.2-1), rust-pcre2 (= 0.2.3-1), rust-pcre2-sys (= 0.2.5-1), rust-regex-automata (= 0.1.8-2), rust-regex (= 1.3.9-1), rust-regex-syntax (= 0.6.25-1), rust-ryu (= 1.0.2-1), rust-same-file (= 1.0.6-1), rust-serde (= 1.0.130-1), rust-serde-json (= 1.0.41-1), rust-strsim (= 0.9.3-1), rust-termcolor (= 1.1.0-1), rust-textwrap (= 0.11.0-1build1), rust-thread-local (= 1.1.3-3), rust-unicode-width (= 0.1.8-1), rust-walkdir (= 2.3.1-1), rustc (= 1.53.0+dfsg1+llvm-4ubuntu1)
./md5sums0000644000000000000000000000060314113505654011234 0ustar  rootroot82cc2a1fdf7721bd88911cb638a71e22  usr/bin/rg
536668671f52e8e32f9f607aa4b70816  usr/share/doc/ripgrep/TODO.Debian
d613394d6c66ef2cd2ee0ed5fafcaa32  usr/share/doc/ripgrep/changelog.Debian.gz
f78c0f9ef0e769e9929d5e942a11ca02  usr/share/doc/ripgrep/copyright
ef4cf19ff3297de685a0ab8e6a12ecf1  usr/share/man/man1/rg.1.gz
7d274f4cd16b398d1b72173ebbf5c492  usr/share/zsh/vendor-completions/_rg
```

Reply:
Yeah @schopin-pro I remember seeing the same on my initial more thorough analysis - which along the discussion here got us to the rules we had, it even is in the resolved conversations of this PR :-). But this time missed that little _x - thanks!

I guess the question will have to be - do we:
a) change it back to X-Cargo-Built-Using?
b) do we file a bug against dh-cargo/rustc which needs to be resolved as part of the toolchain MIR to ensure this info lands in `Built-using` for Ubuntu *1.

@vorlonofportland - would you prefer and support such a post build mangling in the ubuntu builds to make normal Built-Using work?

*1 implementation probably needs to be in a different component since the packages refer to the field in their d/control it can't be fixed in dh-cargo but would need to be a post build mangling of the meta data. Probably as part of pkgbinarymangler or such.

Reply:
Oh BTW @schopin-pro that bug for Built-using (if we create it) would go along another one for the already discussed need to make our dh-cargo to allow for proper .lock files.

Reply:
I had a look at the Debian policy regarding `Built-Using`, and it is stated there that the purpose of the field is strictly for licensing issues. See https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=921284 which is where the issue was raised for golang. There's been some inquiries from the Go side regarding migrating X-Cargo-Built-Using towards a more neutral field that could then be used by dh-golang as well, I'll keep an eye on the situation.

Reply:
> ...
> I'll keep an eye on the situation.

Thank you!